### PR TITLE
struct_target_features: reduce serialized rmeta

### PR DESCRIPTION
**Pull request by veluca93**
Reduce the amount of serialized information in rmeta. This should fix the binary size regression in #127537 and *may* also fix the speed regression.

r? compiler-errors

Tracking issue: #129107

### CODE REVIEW

File: compiler/rustc_metadata/src/rmeta/encoder.rs [hunk 1]
```diff
@@ -1393,7 +1393,10 @@ impl<'a, 'tcx> EncodeContext<'a, 'tcx> {
                 record!(self.tables.codegen_fn_attrs[def_id] <- self.tcx.codegen_fn_attrs(def_id));
             }
             if def_kind.has_struct_target_features() {
-                record_array!(self.tables.struct_target_features[def_id] <- self.tcx.struct_target_features(def_id));
+                let struct_target_features = self.tcx.struct_target_features(def_id);
+                if !struct_target_features.is_empty() {
```
Comment (compiler-errors) [inline compiler/rustc_metadata/src/rmeta/encoder.rs:1397]: This is redundant. `record_defaulted_array` already checks if it's empty.

**Comment by Kobzol** [comment]
@bors r-

We think that the original PR might have introduced perf. instability, so we might have to revert it (https://github.com/rust-lang/rust/pull/129854). Let's not add additional stuff on top of it until we resolve the perf. results instability (the perf. result reported above is incomplete anyway, due to our current perf. collector problems).

**Pull request closed without merging**